### ISSUE TITLE
Consumer Process for Discovery

### ISSUE THREAD

**Issue by mmccool** [w3c/wot-discovery#272]
We need a clearer description of what the consumer has to do to implement discovery.  This should include:
- Managing and returning multiple introduction URIs
- Following multiple introduction URIs and merging multiple exploration results into a single *set* of TDs
- If Directories are encountered, optionally automatically following "listing" affordances and merging results... this may or may not be a good idea (e.g. the result list could be large, depending on the context), but the *option* to do it would be convenient in some contexts, e.g. Smart Home using a local hub for discovery
- NOT automatically following external links, e.g. TD Links support federated directories... which implies that if directories are discovered they may need to be included in the output from the discovery process, not elided
- Requiring that consumers that support discovery MUST support directories, at least to the point of using the (mandatory) "listing" affordance
- Consumers having the option to limit use of introduction mechanisms to those that limit scope to the LAN or to nearby devices (e.g. Bluetooth beacons, line of sight QR codes, etc).

Please comment below on the above or any other points that should be included.  The intention here is to capture the process as we intended, not to add significant NEW requirements or functionality.

**Comment by mmccool** [comment]
I think the normative parts here are:
- Consumers MUST support multiple introduction results and multiple TD results
- Consumers MUST support at least "listing" affordance of any Directories TDs encountered
- Consumers MAY automatically expand "listing" affordances and add TDs to results
- Consumers MUST have a way to disable automatic expansion of discovery results if it is inappropriate
- Consumers MAY limit discovery to local context, e.g. to network segment or to proximate things, by limiting the introduction mechanisms used
- Consumers MAY use a limited subset of the introduction mechanisms

**Comment by mmccool** [comment]
- This could all be considerably simplified if we dropped the automatic expansion of directory contents.  However, I think this was our original intention so that's what I'm trying to capture here, but making it optional so it does not become burdensome in cases where the directories have many TDs in them.
- Need to consider how to handle queries.  In the scripting API call, talked about making JSONPath (or something) a mandatory part of the discovery API, which would make it transparent whether it was supported remotely (if the directory supports it) or locally (if the directory does not and post-filtering is required).  Unfortunately JSONPath is not normative so we run into the same problem we had with making it normative on directories so for now I plan to not mention this or make it a requirement.  We should work on this in Scripting and wait for JSON Path to get finalized and deal with it in the next round...

**Comment by mmccool** [comment]
My most recent thinking on this is to keep the requirements on Consumers as simple as possible.  In particular, the more I think about it, the more problems "automatic" expansion causes, so I think that expansion should be under application control and the discovery process should just return the TDs it gets after following all Introduction links, no more.  The application can then choose to pull down multiple TDs from any ThingDirectory TDs.

I am also assuming that the ONLY mandatory affordance in ThingDirectories will be "listing" (@farshidtz is working in parallel on that) allowing implementation of small static (read-only) "directories" for simple multiple-endpoint devices (e.g. power strips), solving the issue of some Intro types only support one URL and therefore only one TD.

My goal is a clear process that ALL consumers can easily implement (assuming they implement an HTTP client).

**Comment by mmccool** [comment]
To be clear, Architecture only has the following assertion at this point, which makes Discovery optional:
* arch-discovery-should-use-standard: Whenever possible, the Discovery mechanisms defined in [[!WOT-DISCOVERY]] SHOULD be used to distribute TDs.

So Discovery is NOT mandatory for all Consumers.

**Comment by k-toumura** [comment]
Instead of the term "Consumer", how about to use another term, such as Discovery Client, Discovery User, or Discoverer, etc.?

I don't think that users of Discovery are necessarily "Consumer" in the sense of WoT.  For example, in [my previous implementation of Node-RED-WoT-Discovery integration](https://github.com/w3c/wot-testing/blob/main/events/2020.09.Online/prototypes/Hitachi/2020-10-01-nodered-mdns.pdf), The Auto-population Server searches TDs by mDNS, then create and register the corresponding nodes to Catalogue Server.  In this case, even though the Auto-population Server is a user of Discovery, but it is not a WoT Consumer (= "An entity that can process WoT Thing Descriptions *and interact with Things*")

**Comment by mmccool** [comment]
Discussion in Scripting API call:
- Need to clarify use cases
- Multiple minimum requirements: URL (Direct), Automatic (e.g. DNS-SD, etc)
- Change in goals: list of TDs -> application choice; more complex API needed
- Original plan was to use server-side queries to filter; unfortunately...
- Complexity: looks like LDAP, LDAP is complicated; general "tree expansion"
- Another idea would be to do automatic breadth-first expansion
- Could also have multiple "Discovery Objects" in the script API; e.g. one Direct, one Automatic
- There are also Thing Links to dereference
- And composed models; and now also composed TDs.  Composed TDs could be treated like Thing Links
- Also mini-directories for hosts with multiple static things
- For one-URL introductions, e.g. .well-known, and the use cases of a power outlet modelled as a composed TD, do we point only at the parent TD, or to a mini-directory?  Should be specified which is the recommended option.  And when do you use one or other?  If you use the first, and the "element" TDs do not exist in the first set of results, then they have to be "expanded".

**Comment by mmccool** [comment]
See also https://github.com/w3c/wot-scripting-api/issues/363 - how to deal with composite TDs, relates to how we deal with Thing Links.

**Comment by zolkis** [comment]
Looking from an application script's (developer's) point of view, we have the following stakeholders:
- script (client app)
- (client) user agent (e.g. node-wot)
- WoT discovery service (directory), or
- Things existing on the network (local or not, addressable by URLs).

Possibly one more, a _solution_, which might include the design (and hence extra heuristics) of a directory service, how TDs are organized in it (by URLs), and also including the script that is run by a generic WoT implementation.

The basic use cases of discovery from a script are 
- "give me TDs from this directory", 
- "give me the TD from this URL", 
- "give me TDs that match these filters".

Since filtering is not implemented (yet) on server side, the argument was that if an implementation (node-wot) does filtering locally, then it might not use heuristics the applications/solutions/scripts might have, and might block in too big directories, eclipsing smaller ones, etc. As a remedy, the suggestion was to expose navigating between directories, links, etc by the client script, hence expose an API for navigating. That sounds like LDAP/structure/DOM parsing (with similar, or actually even bigger complexity). IMHO it is not realistic to expose such complexity to scripts for the reasons above, unless we can considerably reduce that complexity.

I'd argue that in the cases when there is extra heuristics about how TDs are organized and exposed (e.g. inside a solution), that intrinsic knowledge can be reflected by the script "knowing" what URLs or directories to use, or from where to fetch, i.e. there is no actual need for a generic navigation and introspection API, since the heuristics can be built in the script. That would allow to specify a generic scripting API with the use cases above, and without the structure-parsing API.

Then there are the privacy related concerns (e.g. following ThingLinks cross origin etc). Usually these are handled by the (trusted) runtime/implementation (e.g. using pop-up permissions, or provisioned permissions), and not by the scripts (which are not trusted).

In summary, I would try to encapsulate these in the implementations, as long as possible.

**Comment by mmccool** [comment]
I think we are done with this issue now, with the new "Discoverer" section, except for some cleanup around terminology (which there are other issues for).
